readthedocs/api/v2/permissions.py

```diff
@@ -65,3 +67,37 @@ def has_permission(self, request, view):
             .exists()
         )
         return has_access
+
+
+class TokenKeyParser(KeyParser):
+
+    """
+    Custom key parser to use ``Token {TOKEN}`` as format.
+
+    This is the same format we use in API V3 for auth/authz.
+    """
+
+    keyword = "Token"
+
+
+class HasBuildAPIKey(BaseHasAPIKey):
+
+    """
+    Custom permission to inject the build API key into the request.
+
+    This avoids having to parse the key again on each view.
+    The key is injected in the ``request.build_api_key`` attribute
+    only if it's valid, otherwise it's set to ``None``.
+    """
+
+    model = BuildAPIKey
+    key_parser = TokenKeyParser()
+
+    def has_permission(self, request, view):
+        build_api_key = None
+        has_permission = super().has_permission(request, view)
```

Comment:
We rely on the permission class from the package to make sure that the API key given is valid, we just override it to attach the key to the request, otherwise we need to parse the  key and get it from the db every time we want to use from a view.

https://github.com/readthedocs/readthedocs.org/blob/f261ecc69884421332c93f007532ef63ada35f36/readthedocs/api/v2/permissions.py#L85-L91

Reply:
Gotcha, so it's basically just checking that the key exists? Not checking it against any specific user or anything.

Reply:
yeah, api keys aren't attached to a user, authorization for a given project is done via the queryset.